blocks/iterpool.go

```diff
@@ -0,0 +1,47 @@
+package blocks
+
+import (
+	"errors"
+	"runtime"
+	"sync"
+
+	"github.com/bsm/go-sparkey"
+)
+
+var errOpenFailed = errors.New("failed to open block iter")
+
+// iterPool is a sync.Pool for *sparkey.HashIter objects. Since each one
+// represents a sparkey buffer in cgo, it's important to reuse them where
+// possible. Since cgo file descriptors don't get closed on GC, we use a
+// finalizer to do that.
+type iterPool struct {
+	*sync.Pool
+}
+
+// newIterPool creates a sync.Pool for *sparkey.HashIter objects.
+//
+// The new func returns nil if the sparkey.HashIter couldn't be created.
+func newIterPool(sparkeyReader *sparkey.HashReader) iterPool {
+	return iterPool{
+		Pool: &sync.Pool{
+			New: func() interface{} {
+				iter, err := sparkeyReader.Iterator()
+				if err != nil {
+					return nil
+				}
+
+				runtime.SetFinalizer(iter, (*sparkey.HashIter).Close)
+				return iter
+			},
+		},
+	}
+}
+
+func (ip iterPool) getIter() (*sparkey.HashIter, error) {
```

Comment:
I'd just ditch your `New` above and put all the construction logic down here. That way you get to return more useful `error` values straight from the source, and you can spare my eyes 5-deep indentation 🙈